src/preprocessing/passes/ackermann.cpp

```diff
@@ -187,9 +188,110 @@ void collectFunctionsAndLemmas(FunctionToArgsMap& fun_to_args,
 
 /* -------------------------------------------------------------------------- */
 
+/* Given a minimum capacity for an uninterpreted sort, return the size of the
+ * new BV type */
+size_t getBVSkolemSize(size_t capacity) { return (size_t)log2(capacity) + 1; }
+
+/* Given the lowest capacity requirements for each uninterpreted sorts, assign
+ * a sufficient bit-vector size. Get the converting map */
+void collectUSortsToBV(const vector<TNode>& terms,
+                       USortToBVSizeMap& usortCardinality,
+                       SubstitutionMap& sortsToSkolem)
+{
+  NodeManager* nm = NodeManager::currentNM();
+
+  for (TNode term : terms)
+  {
+    TypeNode type = term.getType();
+    size_t size = getBVSkolemSize(usortCardinality[type]);
+    Node skolem = nm->mkSkolem(
+        "BVSKOLEM$$",
+        nm->mkBitVectorType(size),
+        "a variable created by the ackermannization "
+        "preprocessing pass, representing a term with uninterpreted sort.");
+    sortsToSkolem.addSubstitution(term, skolem);
+  }
+}
+
+/* This function returns the list of terms with uninterpreted sort in formula
+ * represented by assertions. We use a BFS to get all terms without
+ * duplications. */
+std::vector<TNode> getVarsWithUS(AssertionPipeline* assertions)
+{
+  TNodeSet seen;
+  std::vector<TNode> res;
+
+  for (Node& assertion : assertions->ref())
+  {
+    std::unordered_set<TNode, TNodeHashFunction> vars;
+    expr::getVariables(assertion, vars);
+
+    for (const TNode& var : vars)
+    {
+      if (var.getType().isSort() && seen.find(var) == seen.end())
+      {
+        seen.insert(var);
```

Comment:
vars is a set for single assertion, while seen is over all assertions.

Reply:
Ah yes, that is a good point. How about just returning the set instead of the vector? We just need a collection that we can iterate over, so there is no need to compute two representations. We can also reuse the set that we pass to `expr::getVariables()` since that function does not clear the set, so this whole method becomes something like:

```
  TNodeSet vars;
  for (const Node& assertion : assertions->ref())
  {
    expr::getVariables(assertion, vars);
  }
  return vars;
```

Reply:
Good point to change the vector to set. But is this a bit hack? Actually, it maybe confusing about the name of the function. Just as my experiences, usually, getXXX() should return the value we want to get.
But this maybe a historical thing, I will just change it as you suggested.

Reply:
Maybe I am misunderstanding but you could argue that the value that want to get is a set of variables with uninterpreted sorts, so the name is not wrong :) `getVarsWithUSorts` doesn't say anything about whether the variables are in a set or in a vector, IMHO.

Reply:
Oh, and if I want to return a list with only U sort terms, it is better to have separate them.